---
title: "Ribosome profiling analysis pipeline"
author: "Yating Liu"
date: "9/9/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/path/to/root/directory")
library(tidyverse)
library(edgeR)
library(biomaRt)
library(gridExtra)
library(genefilter)

save_figs <- F
options(stringsAsFactors = FALSE)
source("/path/to/ribo_profiling_functions.R")
```

## Specify the relative output directory and create necessary folders
```{r}
## global variable: relative path of the output folder 
OUTPUT <- "exp1_results"
## global variable: group of samples 
SAMPLES <- c("RP3_mock", "RP9_mock", "RP3_12hpi", "RP9_12hpi")
make_output_directories(OUTPUT)

## check dependency: need a samples.csv file to describe replicates, treatements and etc for each sample in the current root dir
if (!file.exists("samples.csv")) {
  stop("samples.csv doesn't exist!")
}
```

## make DGE objects from Ribo-seq and RNA-seq gene counts

> Use htseq-count to count CDS and copy the results folder into OUTPUT, name the folder as htseq_count_cds

```{r}
if (!dir.exists(file.path(OUTPUT, "htseq_count_cds"))) {
  stop("HTseq-count results for CDS doesn't exist!")
}

# Make DGE object for all samples
make_dge(file.path(OUTPUT, "htseq_count_cds"), "", region = "cds")

```

## QC

### reproducibility
```{r}

all_cpm <- read_csv(file.path(OUTPUT, "reports","cpm_cds.csv"))

generate_reprod_plot(all_cpm, "riboseq")
generate_reprod_plot(all_cpm, "rnaseq")
```

### MDS

```{r}
dge_all <- readRDS(file.path(OUTPUT, "objs", "dge_cds_protein_coding.rds"))


generate_mds(dge_all, "riboseq")
generate_mds(dge_all, "rnaseq")

```

## find the threshold to cut the low expressed genes
```{r}
require(genefilter)

replications <- list(c("RP3_mock", "RP9_mock"), c("RP3_12hpi", "RP9_12hpi"))
plot_CV_vs_CPM(all_cpm, replications, "rnaseq")
plot_CV_vs_CPM(all_cpm, replications, "riboseq")

```

### reproducibility after filter
Use the CV vs CPM plots to find the filter threshold:
log2 cpm > 2
# adjust threshold and numRep
```{r}
filter_dge(dge_all,"cds", 2, 2)
dge_all_filt <- readRDS("exp1_results/objs/dge_cds_filt_norm.rds")
all_cpm_filt <- read_csv("exp1_results/reports/cpm_cds_filt.csv")
generate_reprod_plot(all_cpm_filt, "rnaseq")
generate_reprod_plot(all_cpm_filt, "riboseq")
```

## Check if Ribosome profiling work

### CDS TE distribution
```{r}
combined_table <- get_table_for_samples(SAMPLES, all_cpm_filt) 
  
for (sample in SAMPLES) {
  dens_plot <- ggplot(combined_table, aes_string(x=paste0(sample, "_te"))) +
    geom_density() +
    xlim(-5, 5)
  if (save_figs) ggsave(file.path(file.path(OUTPUT, "reports", "figs"), paste0(sample, "_te_cds.pdf")), plot = dens_plot)
  print(dens_plot)
}
```



#### Generate cds GTF, utr3 GTF and utr5 GTF 

```{r, eval=F}
ensembl = useMart(biomart = "ENSEMBL_MART_ENSEMBL", version = "Ensembl Genes 90", dataset = "csabaeus_gene_ensembl", host = "http://aug2017.archive.ensembl.org")

filtered_genes_ens <- getBM(attributes=c('ensembl_gene_id','ensembl_transcript_id', 'chromosome_name', 'strand', 'external_gene_name', '5_utr_start', '5_utr_end', '3_utr_start', '3_utr_end', 'genomic_coding_start', 'genomic_coding_end'),
                 filters = 'ensembl_gene_id', values = all_cpm_filt$ensembl_gene_id, mart = ensembl)

cds_regions <- filtered_genes_ens %>% 
  filter(!is.na(genomic_coding_start) & !is.na(genomic_coding_end)) %>% 
  transmute(chr = chromosome_name, source="protein_coding", method="CDS", start=genomic_coding_start, end=genomic_coding_end, score=".", 
            strand=case_when(strand == -1 ~ "-",
                             strand == 1 ~ "+"), 
            phase=".",
            group = paste0('gene_id "', ensembl_gene_id, '"; ', 'transcript_id "', ensembl_transcript_id, '"; ', ' gene_name "', external_gene_name, '";')) %>% 
  arrange(chr, start, end)
write.table(cds_regions, file.path(OUTPUT, "cds_regions.gtf"), col.names = F, row.names = F, quote = F, sep = "\t")

utr5_annotation <- filtered_genes_ens %>% 
  filter(!is.na(`5_utr_start`) & !is.na(`5_utr_end`)) %>% 
  transmute(chr = chromosome_name, source="protein_coding", method="UTR5", start=`5_utr_start`, end=`5_utr_end`, score=".", 
            strand=case_when(strand == -1 ~ "-",
                             strand == 1 ~ "+"), 
            phase=".",
            group = paste0('gene_id "', ensembl_gene_id, '"; ', 'transcript_id "', ensembl_transcript_id, '"; ', ' gene_name "', external_gene_name, '";')) %>% 
  arrange(chr, start, end)
write.table(utr5_annotation,file.path(OUTPUT, "utr5_annotation.gtf"), col.names = F, row.names = F, quote = F, sep = "\t")


utr3_annotation <- filtered_genes_ens %>% 
  filter(!is.na(`3_utr_start`) & !is.na(`3_utr_end`)) %>% 
  transmute(chr = chromosome_name, source="protein_coding", method="UTR3", start=`3_utr_start`, end=`3_utr_end`, score=".", 
            strand=case_when(strand == -1 ~ "-",
                             strand == 1 ~ "+"), 
            phase=".",
            group = paste0('gene_id "', ensembl_gene_id, '"; ', 'transcript_id "', ensembl_transcript_id, '"; ', ' gene_name "', external_gene_name, '";'))
write.table(utr3_annotation, file.path(OUTPUT, "utr3_annotation.gtf"), col.names = F, row.names = F, quote = F, sep = "\t")

```

#### Removed the regions overlapped with CDS, and removed the regions overlapped with repeats, then use htseq-count to count reads within 3' or 5'

> download appropriate repeat masker track in GTF format from UCSC and then edit the chromosome names (by removing "chr") to be consist with Ensembl

e.g. for hg19: hg19RepMasker_removed_chr.gtf

> generate cds GTF, utr3 GTF and utr5 GTF. Use bedtools subtract to removed overlapped regions for utr3 and utr5

* bedtools subtract -a utr3_annotation.gtf -b cds_regions.gtf > utr3_annotation_removed_cds.gtf
* bedtools subtract -a utr3_annotation_removed_cds.gtf -b hg19RepMasker_removed_chr.gtf > utr3_annotation_cleaned.gtf
* bedtools subtract -a utr5_annotation.gtf -b cds_regions.gtf > utr5_annotation_removed_cds.gtf
* bedtools subtract -a utr5_annotation_removed_cds.gtf -b hg19RepMasker_removed_chr.gtf > utr5_annotation_cleaned.gtf

> Use htseq-count or featureCount to count reads in utr3, utr5 with the cleaned annotation file with strict mode

Download the htseq-count result folders to OUTPUT, name the folder as htseq_count_utr3_strict and htseq_count_utr5_strict



### TE of CDS, 5' and 3' UTR
```{r}
if (!dir.exists(file.path(OUTPUT, "htseq_count_utr3_strict"))) {
  stop("HTseq-count results for 3' UTR doesn't exist!")
}
if (!dir.exists(file.path(OUTPUT, "htseq_count_utr5_strict"))) {
  stop("HTseq-count results for CDS doesn't exist!")
}

## make dge for utr5 and utr3 counts 
make_dge(file.path(OUTPUT, "htseq_count_utr3_strict"), "", region = "utr3")
make_dge(file.path(OUTPUT, "htseq_count_utr5_strict"), "", region = "utr5")
dge_utr5 <- readRDS(file.path(OUTPUT,"objs", "dge_utr5_protein_coding.rds"))
dge_utr3 <- readRDS(file.path(OUTPUT,"objs", "dge_utr3_protein_coding.rds"))

filter_dge(dge_cds_st, "cds_st", 2, 2)
filter_dge(dge_utr5, "utr5", 2, 2)
filter_dge(dge_utr3, "utr3", 2, 2)

cpm_utr5 <- read_csv(file.path(OUTPUT, "reports", "cpm_utr5_filt.csv"))
combined_table_utr5 <- get_table_for_samples(SAMPLES, cpm_utr5)
cpm_utr3 <- read_csv(file.path(OUTPUT, "reports", "cpm_utr3_filt.csv")) 
combined_table_utr3 <- get_table_for_samples(SAMPLES, cpm_utr3)

plot_individual_density(combined_table_utr3, "utr3")
plot_individual_density(combined_table_utr5, "utr5")

te_colnames <- paste0(SAMPLES, "_te")
kept_colnames <- append(c("ensembl_gene_id"), te_colnames)
combined_table_all_te <- left_join(combined_table_utr5[,kept_colnames], combined_table_utr3[,kept_colnames], by = "ensembl_gene_id", suffix=c("_utr5", "_utr3"))
combined_table_all_te <- left_join(combined_table_all_te, combined_table[,kept_colnames], by = "ensembl_gene_id") 
combined_table_all_te <- combined_table_all_te %>% 
  dplyr::rename(RP3_mock_te_cds = RP3_mock_te,
         RP3_12hpi_te_cds = RP3_12hpi_te,
         RP9_mock_te_cds = RP9_mock_te,
         RP9_12hpi_te_cds = RP9_12hpi_te)

combined_table_all_te_melt <- gather(combined_table_all_te, key="sample", value = "TE", -ensembl_gene_id) 
combined_table_all_te_melt$region <- sapply(str_split(combined_table_all_te_melt$sample, "_"), `[[`, 4)
combined_table_all_te_melt$group <- str_remove(combined_table_all_te_melt$sample, paste0("_te_", combined_table_all_te_melt$region))
write.csv(combined_table_all_te_melt, file.path(file.path(OUTPUT, "reports"), paste0("combined_table_all_te_melt.csv", ".csv")))
                                            
for (sam in SAMPLES) {
  dens_plot <- ggplot(subset(combined_table_all_te_melt, group == sam), aes(x=TE, group = region, fill = region)) +
    geom_density(alpha = 0.3) +
    xlim(-5, 5) +
    labs(title = sam)
  if (save_figs) ggsave(file.path(file.path(OUTPUT, "reports", "figs"), paste0(sam, "_te_cds_utr.pdf")), plot = dens_plot)
  print(dens_plot)
}

```

### Scatterplot of RNAseq vs Riboseq 
```{r}
for (sample in SAMPLES) {
  sub_table <- data.frame(rnaseq = combined_table[[paste0(sample, "_rnaseq")]],
                          riboseq = combined_table[[paste0(sample, "_rf")]])
  m <- lm(riboseq ~ rnaseq, sub_table)
  r2 = paste0("italic(R) ^ 2 == ", format(summary(m)$r.squared, digits = 3))
  
  pt <- ggplot(sub_table, aes(x=rnaseq, y = riboseq)) +
    geom_point(alpha = 0.6, color = "grey50") +
    annotate("text", x = 0, y = 15, label = r2, parse = TRUE) +
    labs(x=paste0(sample, " RNA-seq (log2 CPM)"), y = paste0(sample, " RF (log2 CPM)")) +
    theme_bw()
  if (save_figs) ggsave(file.path(file.path(OUTPUT, "reports", "figs"), paste0(sample, "_rnaseq_vs_riboseq_scatter.pdf")), plot = pt)
  print(pt)
}
```

### Get P site information

> On HTCF, copy Transcriptome alignment BAM files to a folder transcriptome_bams. Rename the filenames to sample names

* SK222_Aligned.toTranscriptome.out.bam -> M_0_rf.bam
* SK391_Aligned.toTranscriptome.out.bam -> M_24_rf.bam
* SK393_Aligned.toTranscriptome.out.bam -> GM_0_rf.bam
* SK394_Aligned.toTranscriptome.out.bam -> GM_24_rf.bam


```{r}
library(riboWaltz)
require(riboWaltz)
if (!file.exists(file.path(OUTPUT, "objs", "riboWaltz.RData"))) {
  annotation <- create_annotation(gtfpath = "/path/to/annotation/file/chlSab2.ncbiRefSeq.gtf")
  reads_list <- bamtolist(bamfolder = "/path/to/transcriptome_bams", annotation = annotation)
  reads_list_filter <- length_filter(data = reads_list, length_filter_mode = "custom", length_filter_vector = 25:33)
  psite_offset <- psite(reads_list_filter, flanking = 6, extremity = "auto")
  reads_psite_list <- psite_info(reads_list_filter, psite_offset)
  save(reads_list, reads_list_filter, annotation, psite_offset, reads_psite_list, file = "exp1_results/objs/riboWaltz.RData")
}
load(file.path(OUTPUT, "objs", "riboWaltz.RData"))

```
### P-site frames
```{r}
for (sample in SAMPLES) {
  example_frames <- frame_psite(reads_psite_list,sample = paste0(sample, "_rf"), region = "all")

  pt <- example_frames[["plot"]] +
    labs(title = sample)
  if (save_figs) ggsave(filename = paste0(OUTPUT, "/reports", "/figs/", sample, "_rf_frame_psite.png"), plot = pt)
  print(pt)
}
```

### P-site frames - Heatmap
```{r}
example_frames_heat <- frame_psite_length(reads_psite_list, region = "all")
                                          
pt <- example_frames_heat[["plot"]]
pt <- pt + theme(strip.text.y = element_text(size=10))
if (save_figs) ggsave(filename = paste0(OUTPUT, "/reports", "/figs/", "frame_psite_heatmap.pdf"), plot = pt)
print(pt)

```

### P-site metaprofile
```{r}
for (sample in SAMPLES) {
  example_metaprofile <- metaprofile_psite(reads_psite_list, annotation, sample = paste0(sample, "_rf"), utr5l = 25, cdsl =50, utr3l = 25, plot_title = "sample.transcript")
  pt <- example_metaprofile[[paste0("plot_",sample, "_rf")]]
  pt <- pt + theme(plot.title = element_text(size = 20), strip.text.x = element_text(size=15))
  if (save_figs) ggsave(file.path(file.path(OUTPUT, "reports", "figs"), paste0(sample, "_rf_metaprofile_psite.png")), plot = pt, width = 10, height = 6)
  print(pt)
  
}
```


### metagene analysis

On HTCF, copy alignment bam files (*_Aligned.sortedByCoord.out.bam) to a folder metagene_bams and rename the files to samples names. Sort and index.

```{r}
require(metagene)
if (!file.exists(file.path(OUTPUT, "objs","mg.rds"))) {
  bam_dir <- "C:/Users/nlee1/Documents/lab_data/Vero_Yating/Vero_12hpi/exp1_results/metagene_bams"
  bam_files <- list.files(bam_dir, pattern = ".bam$")
  CHUNK_LENGTH <- 1000
  # get the region TSS - TSS+1000 and region ~1000 - stop codon for riboseq genes
  #ensembl = useMart(biomart = "ENSEMBL_MART_ENSEMBL", version = "Ensembl Genes 75", dataset = "hsapiens_gene_ensembl", host = "http://feb2014.arc12hpie.ensembl.org")
  ensembl = useMart(biomart = "ENSEMBL_MART_ENSEMBL", version = "Ensembl Genes 90", dataset = "csabaeus_gene_ensembl", host = "http://aug2017.archive.ensembl.org")
  cds_region <- getBM(attributes=c('ensembl_gene_id','ensembl_transcript_id', 'chromosome_name', 'strand', 'external_gene_name', 'genomic_coding_start', 'genomic_coding_end', 'cds_length'), filters = 'ensembl_gene_id', values = all_cpm_filt$ensembl_gene_id, mart = ensembl)
  
  ## keep the first and last cds regions for each transcripts
  cds_region_filtered_with_codon <- cds_region %>% 
    group_by(ensembl_transcript_id) %>% 
    filter(!is.na(genomic_coding_start) & !is.na(genomic_coding_end)) %>% 
    filter(row_number() == 1 | row_number() == n()) %>% 
    mutate(cds_chunk_length = genomic_coding_end - genomic_coding_start,
           strand = case_when(strand == 1 ~ "+",
                              strand == -1 ~ "-"),
           score = ".") 
  
  
  start_codon_downstream <- cds_region_filtered_with_codon %>% 
    mutate(start = case_when(strand == "+" ~ min(genomic_coding_start),
                             strand == "-" ~ max(genomic_coding_start)),
           end = start + CHUNK_LENGTH,
           cds_chunk_end = case_when(strand == "+" ~ min(genomic_coding_end),
                                     strand == "-" ~ max(genomic_coding_end))) %>% 
    #filter(end <= cds_chunk_end) %>%
    dplyr::select(chromosome_name, start, end, ensembl_transcript_id, score, strand) %>% 
    distinct()
  write_tsv(start_codon_downstream, file.path(OUTPUT, "reports", "startCodon.bed"), col_names = F)
  
  stop_codon_upstream <- cds_region_filtered_with_codon %>% 
    mutate(end = case_when(strand == "+" ~ max(genomic_coding_end),
                           strand == "-" ~ min(genomic_coding_end)),
           start = end - CHUNK_LENGTH,
           cds_chunk_start = case_when(strand == "+" ~ max(genomic_coding_start),
                                     strand == "-" ~ min(genomic_coding_start))) %>%
    #filter(start >= cds_chunk_start) %>%
    dplyr::select(chromosome_name, start, end, ensembl_transcript_id,  score, strand) %>% 
    distinct()
  write_tsv(stop_codon_upstream, file.path(OUTPUT, "reports", "stopCodon.bed"), col_names = F)
  
  
  regions <- c(file.path(OUTPUT, "reports", "startCodon.bed"), file.path(OUTPUT, "reports", "stopCodon.bed"))
  mg <- metagene$new(regions = regions, 
                     bam_files = file.path(bam_dir, bam_files))
  design <- data.frame(samples = file.path(bam_dir, bam_files), time_0 = case_when(str_detect(bam_files, "0")== TRUE ~ 1, TRUE ~ 0), time_24 = case_when(str_detect(bam_files, "24")== TRUE ~ 1, TRUE ~ 0))
  mg$produce_table(design = design, normalization = "RPM", bin_count = CHUNK_LENGTH, flip = T)
  mg$produce_data_frame()
  saveRDS(mg, file.path(OUTPUT, "objs","mg.rds"))
}
```

### plot metagene
```{r}
mg <- readRDS(file.path(OUTPUT, "objs","mg.rds"))
mg_df <- mg$get_data_frame()
ggplot(subset(mg_df, region=="startCodon"), aes(x=bin, y=value, group = group, color = group)) +
  geom_line() +
  labs(title = "1000nt downstream start codon")

ggplot(subset(mg_df, region=="stopCodon"), aes(x=bin, y=value, group = group, color = group)) +
  geom_line() +
  labs(title = "1000nt upstream stop codon")


```

## Differential expression analysis
```{r}
de_analysis(dge_all_filt)
```

### heatmap 
```{r}
require(RColorBrewer)
require(gplots)

de_rnaseq <- read_csv(file.path(OUTPUT, "reports", "de_genes", "time_24_vs_0_rnaseq.csv"))
de_riboseq <- read_csv(file.path(OUTPUT, "reports", "de_genes", "time_24_vs_0_riboseq.csv"))
de_te <- read_csv(file.path(OUTPUT, "reports", "de_genes", "time_24_vs_0_te.csv"))

union_de_genes <- union(union(de_riboseq$ensembl_gene_id, de_rnaseq$ensembl_gene_id), de_te$ensembl_gene_id)

combined_fc <- data.frame(ensembl_gene_id = union_de_genes) %>% 
  left_join(de_rnaseq[,c("ensembl_gene_id", "logFC")], by = "ensembl_gene_id") %>% 
  left_join(de_riboseq[,c("ensembl_gene_id", "logFC")], by = "ensembl_gene_id", suffix=c("", "_rf")) %>% 
  left_join(de_te[,c("ensembl_gene_id", "logFC")], by = "ensembl_gene_id", suffix=c("", "_te")) %>%
  dplyr::rename(logFC_rnaseq = logFC)

plot_heatmap(combined_fc)

```
